streams-for-raw-video-explainer.md

```diff
@@ -0,0 +1,124 @@
+
+Streams APIs can create coupling between the processing units of the pipeline when objects in the pipe are mutable:
+a processing unit A might pass an object O to another unit B through a stream (say a `WritableStream`) W.
+If A keeps a reference to O, it might mutate O while B is doing processing based on O.
+
+This is especially an issue with objects like `VideoFrame` objects that need to be closed explicitly.
+Taking the previous example, if A decides to close a `VideoFrame` after passing it to W but before B gets it, B will receive a closed `VideoFrame`, which is probably considered a bug.
+If A does not close `VideoFrame`, it is the responsibility of the remaining of the pipeline to close it.
+There is a need to clearly identify who is owning these objects and who is responsible to close these objects at any point in time.
+
+The proposed solution is to transfer ownership of an object to the stream when it gets written or enqueueud to the stream.
+By doing so, the processing unit that enqueues/writes objects will not be able to mutate objects manipulated by the stream and is relieved of the lifetime management of these objects.
+Conversely, processing units take ownership of objects when they receive them from a stream.
+
+Transferring ownership should be opt-in. For that purpose, a new streams type, named 'transfer' in this document,  would be added.
+
+## Example
+
+Below is an example of JavaScript that shows how this can be used.
+The example creates a processing pipe starting with a camera stream and applying two transforms, one for doing a processing for every 30 frame, and one for doing background blur.
+
+```javascript
+// JS transform
+let frameCount = 0;
+const frameCountTransform = new TransformStream({
+  transform: async (videoFrame, controller) => {
+    try {
+      // videoFrame is under the responsibility of the script and must be closed when no longer needed
+      controller.enqueue(videoFrame);
+      // At this point, videoFrame has been transfered within controller.enqueue call. frameCountTransform cannot mutate it.
+      if (!(++frameCount % 30) && frameCountTransform.onEach30Frame)
+          frameCountTransform.onEach30Frame(frameCount);
+    } catch {
+      videoFrame.close();
+    }
+  },
+  readableType: 'transfer',
+  writableType: 'transfer'
+});
+frameCountTransform.onEach30Frame = (frameCount) => {
+  // Do something periodically.
+};
+// Native transform is of type 'transfer'
+const backgroundBlurTransform = new BackgroundBlurTransform();
+
+const cameraStream = await navigator.mediaDevices.getUserMedia({ video : true });
+const videoFrameStream = getReadableStreamFromTrack(cameraStream.getVideoTracks()[0]);
+const blurredVideoFrameStream = videoFrameStream.pipeThrough(backgroundBlurTransform)
+                                                .pipeThrough(frameCountTransform);
+const blurredStream = new MediaStream([getTrackFromReadableStream(blurredVideoFrameStream)]);
+// Make use of blurredStream.
```

Comment:
Use workers part 2. Also, mediacapture-transform is (almost as of this writing) adopted, so I think we can be specific now:
```suggestion
self.onmessage = async ({data: {cameraTrack}}) => {
  const {readable} = new MediaStreamTrackProcessor({cameraTrack});
  const {writable, track} = new VideoTrackGenerator();
  parent.postMessage({track}, [track]);
  await readable.pipeThrough(backgroundBlurTransform).pipeThrough(frameCountTransform).pipeTo(writable);
}

// main.js
const [cameraTrack] = (await navigator.mediaDevices.getUserMedia({video: true}).getVideoTracks();
const worker = new Worker('worker.js');
worker.postMessage({cameraTrack}, [cameraTrack]);

const {data} = await new Promise(r => worker.onmessage);
const blurredStream = new MediaStream([data.track]);
video.srcObject = blurredStream;
```

Reply:
I prefer we keep it abstract from MediaStreamTrackProcessor and we keep the example as small as possible.
I'll remove the call to getUserMedia which is indeed confusing.